### PR TITLE
Add support for rbd ROX PVC mounting

### PR DESCRIPTION
**Pull request by Madhu-1**
if the PVC access mode is ReadOnlyMany or single node read-only, mounting the rbd
device path to the staging path as read-only to avoid the write operation.

Fixes https://github.com/ceph/ceph-csi/issues/725
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
will rebase this after  https://github.com/ceph/ceph-csi/pull/951

File: pkg/rbd/nodeserver.go [hunk 1]
```diff
@@ -386,7 +394,7 @@ func (ns *NodeServer) mountVolumeToStagePath(ctx context.Context, req *csi.NodeS
 	existingFormat, err := diskMounter.GetDiskFormat(devicePath)
 	if err != nil {
 		klog.Errorf(util.Log(ctx, "failed to get disk format for path %s, error: %v"), devicePath, err)
-		return err
+		return readOnly, err
 	}
 
 	if existingFormat == "" && !staticVol {
```
Comment (dillaman) [inline pkg/rbd/nodeserver.go:400]: Do not attempt to create a filesystem on a ROX volume

Comment (Madhu-1) [reply pkg/rbd/nodeserver.go:400]: Thanks for the pointers @dillaman, will make necessary changes

Comment (Madhu-1) [reply pkg/rbd/nodeserver.go]: @dillaman if I don't format it am hitting `cannot mount unformatted disk %s as we are manipulating it in read-only mode` https://github.com/kubernetes/utils/blob/master/mount/mount_linux.go#L353

Comment (dillaman) [reply pkg/rbd/nodeserver.go]: Correct. ROX would need to be a pre-existing disk image (a la static PV) -- or you would need to move all the formatting logic to the `ControllerCreateVolume` call to avoid the potential for races between nodes attempting to format the disk (ROX must always be read-only).

Comment (Madhu-1) [reply pkg/rbd/nodeserver.go]: one use case where the ROX need not be static PV is, cloning a PVC and mounting it as readOnly(ROX).

in `ControllerCreateVolume` we are supposed to create rbd image and map the rbd image and format it ext4 or xfs based on the format type and unmap it one the formatting is completed isn't it?

Comment (dillaman) [reply pkg/rbd/nodeserver.go]: > one use case where the ROX need not be static PV is, cloning a PVC and mounting it as readOnly(ROX).

Cloning/create from snapshot wouldn't be an issue since you would already have a file system on the RBD image.

> in `ControllerCreateVolume` we are supposed to create rbd image and map the rbd image and format it ext4 or xfs based on the format type and unmap it one the formatting is completed isn't it?

`ControllerCreateVolume` just creates the image, correct? I think all mapping and formatting is deferred to the node server plug-in.

Comment (Madhu-1) [reply pkg/rbd/nodeserver.go]: yes, in the current code base mapping and formatting are deferred to the node server.

Comment (dillaman) [reply pkg/rbd/nodeserver.go]: I'm not really suggesting to move it, I am just saying that would be the only thing you could do to safely allow the creation of file systems on ROX volumes. However, I think the ROX use-case should be for static PVs or PVs from existing PVs/snapshots, so no need to re-engineer the code.

Comment (Madhu-1) [reply pkg/rbd/nodeserver.go]: okay let me do some more testing and get back to you :)

**Comment by Madhu-1** [comment]
@dillaman I did some more testing on this one. yes we can mount the rbd image without formatting as the cloned rbd image will have the filesystem on it as you told earlier. however, with current code, the mounting of the rbd image fails
 
```bash
# map rbd images as read-only
rbd map csi-vol-ef7c2743-98f3-11ea-a84a-0242ac110010 --pool=replicapool --read-only

# mount rbd image no the node
sh-4.2# mount -t ext4 -o _netdev,ro,defaults /dev/rbd2 /tmp/test
mount: cannot mount /dev/rbd2 read-only

# dmesg
[11004.179875] EXT4-fs (rbd2): INFO: recovery required on readonly filesystem
[11004.179882] EXT4-fs (rbd2): write access unavailable, cannot proceed (try mounting with noload)
[11059.642337] EXT4-fs (rbd2): INFO: recovery required on readonly filesystem
[11059.642344] EXT4-fs (rbd2): write access unavailable, cannot proceed (try mounting with noload)
[11070.063567] EXT4-fs (rbd2): INFO: recovery required on readonly filesystem
[11070.063568] EXT4-fs (rbd2): write access unavailable, cannot proceed (try mounting with noload)
```
mount rbd image with `noload` or `norecovery` fixes  the issue of mounting. do  we need to make `noload` as default option when mounting  the rbd image?

**Comment by dillaman** [comment]
> mount rbd image with `noload` or `norecovery` fixes the issue of mounting. do we need to make `noload` as default option when mounting the rbd image?

That just sounds like you cloned an in-use/open file system.

**Comment by Madhu-1** [comment]
> > mount rbd image with `noload` or `norecovery` fixes the issue of mounting. do we need to make `noload` as default option when mounting the rbd image?
> 
> That just sounds like you cloned an in-use/open file system.

yes, you are correct, I have cloned PVC which is already attached to the pod. now cloned a PVC which is not attached, working now @dillaman Thanks for the pointer

File: internal/csi-common/nodeserver-default.go [hunk 2]
```diff
@@ -195,3 +195,13 @@ func ConstructMountOptions(mountOptions []string, volCap *csi.VolumeCapability)
 	}
 	return mountOptions
 }
+
+// Contains checks the opt is present in mountOptions
+func Contains(mountOptions []string, opt string) bool {
```
Comment (humblec) [inline internal/csi-common/nodeserver-default.go:200]: This would be better named to MountOptionContains for clarity.

Comment (Madhu-1) [reply internal/csi-common/nodeserver-default.go]: done

**Comment by Madhu-1** [comment]
@humblec @nixpanic PTAL

File: internal/rbd/nodeserver.go [hunk 3]
```diff
@@ -403,10 +412,29 @@ func (ns *NodeServer) mountVolumeToStagePath(ctx context.Context, req *csi.NodeS
...
+		req.VolumeCapability.AccessMode.Mode == csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY {
+		if !csicommon.MountOptionContains(opt, rOnly) {
+			opt = append(opt, rOnly)
+		}
+	}
+	if csicommon.MountOptionContains(opt, rOnly) {
+		readOnly = true
+	}
+
+	if fsType == "xfs" {
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go:433]: This needs a comment. Maybe it is even suitable to be in its own commit (related to volumes with the same uuid).

Comment (Madhu-1) [reply internal/rbd/nodeserver.go:433]: This is just moved from line number 425 to here to keep options in one place

**Comment by Madhu-1** [comment]
> Does the CephFS driver already support ROX as well? If not, support for the CephFS side should be added as well to maintain parity. Also, should the `README.md` table be updated to indicate this support is available?

Added cephfs ROX support in https://github.com/ceph/ceph-csi/pull/1169

**Comment by Madhu-1** [comment]
@humblec @nixpanic resolved merge conflict and updated the Readme

**Comment by humblec** [comment]
> @humblec @nixpanic resolved merge conflict and updated the Readme

Sure.. Thanks.!  Will get on this review soon!

File: e2e/rbd.go [hunk 4]
```diff
@@ -602,6 +604,166 @@ var _ = Describe("RBD", func() {
...
+					err = createPVCAndvalidatePV(f.ClientSet, pvcClone, deployTimeout)
+					if err != nil {
+						Fail(err.Error())
+					}
+					appClone, err := loadApp(appClonePath)
+					if err != nil {
+						Fail(err.Error())
+					}
+
+					totalCount := 4
```
Comment (humblec) [inline e2e/rbd.go:664]: Its better to have a comment over here why this is `4` if there is a specific reason for this number.

Comment (Madhu-1) [reply e2e/rbd.go:664]: There is no reason for it and it is done the same way what we have done in other places also https://github.com/ceph/ceph-csi/blob/master/e2e/rbd.go#L299 https://github.com/ceph/ceph-csi/blob/master/e2e/cephfs.go#L197

**Comment by humblec** [comment]
@Madhu-1 mostly looks good. A minor comment.

File: e2e/rbd.go (hunk 4 above)
Comment (humblec) [reply e2e/rbd.go:664]: Then lets leave it as it is. Thanks.

**Comment by humblec** [comment]
LGTM. Approved.

**Comment by Madhu-1** [comment]
@nixpanic PTAL

**Comment by Madhu-1** [comment]
@nixpanic can you please reivew it.

**Comment by Madhu-1** [comment]
@MergifyIO refresh
